src/main.rs

```diff
@@ -214,13 +244,34 @@ async fn main() {
         });
     }
 
-    // Exit on Ctrl-C (SIGINT) and SIGTERM.
-    let mut term_signal = unix_signal(SignalKind::terminate()).unwrap();
-
-    tokio::select! {
-        _ = signal::ctrl_c() => (),
-        _ = term_signal.recv() => (),
-    };
+    // Initiate graceful shutdown sequence on sig int
+    let mut stream = unix_signal(SignalKind::interrupt()).unwrap();
+
+    stream.recv().await;
+    info!("Got SIGINT, waiting for client connection drain now");
+
+    // Broadcast that client tasks need to finish
+    shutdown_event_tx.send(()).unwrap();
+    // Closes transmitter
+    drop(shutdown_event_tx);
+
+    loop {
```

Comment:
Why are you looping?

Reply:
There should be two events that the receiver gets, the 1st is the event sent by `shutdown_event_tx.send(()).unwrap();` and then the second event would be when all senders are closed and there is an error on the receive channel. made it a loop so that the timeout could be applied to all the operations.